### PR TITLE
BI-13525:alphabetical-search-message-upsert

### PR DESCRIPTION
**Pull request by aokunade**
BI-13525:Upsert message content to search API for Alphabetical search

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/exception/UpsertServiceException.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package uk.gov.companieshouse.alphabeticalcompanysearchconsumer.exception;
+
+public class UpsertServiceException extends RuntimeException  {
```
Comment (tombradley1) [inline src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/exception/UpsertServiceException.java:3]: can't see where this is used. Is this required?

Comment (aokunade) [reply src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/exception/UpsertServiceException.java:3]: Used once in the upsertService method in the UpsertService class.

File: src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/config/Config.java [hunk 2]
```diff
@@ -35,11 +39,19 @@
...
 import uk.gov.companieshouse.service.rest.response.ResponseEntityFactory;
 import uk.gov.companieshouse.stream.ResourceChangedData;
+import java.util.function.Supplier;
 
 @Configuration
 @EnableKafka
 public class Config {
 
+     @Bean
+    public ObjectMapper objectMapper() {
```
Comment (tombradley1) [inline src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/config/Config.java:49]: is this used?

File: src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/config/Config.java [hunk 3]
```diff
@@ -124,5 +136,17 @@ public ConcurrentKafkaListenerContainerFactory<String, ResourceChangedData> kafk
     Logger getLogger() {
         return LoggerFactory.getLogger(NAMESPACE);
     }
+
+    @Bean
+    Supplier<InternalApiClient> internalApiClientSupplier(
```
Comment (tombradley1) [inline src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/config/Config.java]: is this used?

File: src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/service/AlphabeticalIndexUpdaterService.java [hunk 4]
```diff
@@ -31,6 +34,19 @@ public void processMessage(ServiceParameters parameters) {
             ", resource kind " + resourceKind + ", resource URI " + resourceUri + ".",
             getLogMap(message));
 
+            try {
+                upsertService.upsertService(parameters);
```
Comment (tombradley1) [inline src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/service/AlphabeticalIndexUpdaterService.java]: Could we make this more specific? 
Add if statements to ensure that the event.type is correct. 
Take a look in advanced search equivalent for example.

File: src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/config/Config.java (hunk 3 above)
Comment (aokunade) [reply src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/config/Config.java]: yes

File: src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/service/AlphabeticalIndexUpdaterService.java (hunk 4 above)
Comment (aokunade) [reply src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/service/AlphabeticalIndexUpdaterService.java]: will take a look at this now

File: src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/config/Config.java (hunk 3 above)
Comment (aokunade) [reply src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/config/Config.java]: Sorry you're right, I think this is no longer being used. will take it out

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/test/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/service/UpsertServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    upsertService.upsertService(parameters);
+
+     // then
+    verify(privateAlphabeticalCompanySearchHandler).put(
+        eq("/alphabetical-search/companies/" + companyNumber), any(CompanyProfileApi.class));
+    verifyNoInteractions(responseHandler);
+}
+
+@Test
+
```
Comment (tombradley1) [inline src/test/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/service/UpsertServiceTest.java]: extra space

File: src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/util/ApiClientUtils.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+    public static CompanyProfileApi mapMessageToRequest(ServiceParameters parameters) {
+        ResourceChangedData data = parameters.getData();
+        String jsonData = data.getData();
+        
+        try {
+            return objectMapper.readValue(jsonData, CompanyProfileApi.class);
+        } catch (IOException e) {
+            // Handle the exception appropriately
+            e.printStackTrace();
```
Comment (tombradley1) [inline src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/util/ApiClientUtils.java]: Ideally we'd want to stick to using a logger if possible when printing things, and the whole stacktrace could include a lot unnecessary information.

Comment (tombradley1) [reply src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/util/ApiClientUtils.java]: could you instead use the getRootCause in the LoggingUtils?

File: src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/service/AlphabeticalIndexUpdaterService.java [hunk 7]
```diff
@@ -31,6 +36,25 @@ public void processMessage(ServiceParameters parameters) {
             ", resource kind " + resourceKind + ", resource URI " + resourceUri + ".",
             getLogMap(message));
 
+            try {
+                var messageType = message.getEvent().getType();
+                message.getResourceId();
+                switch (messageType) {
```
Comment (tombradley1) [inline src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/service/AlphabeticalIndexUpdaterService.java:42]: This switch statement needs it's indentation fixed.

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/util/ApiClientUtils.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    public static CompanyProfileApi mapMessageToRequest(ServiceParameters parameters) {
+        ResourceChangedData data = parameters.getData();
+        String jsonData = data.getData();        
+        try {
+            return objectMapper.readValue(jsonData, CompanyProfileApi.class);
+        } catch (IOException e) {
+            // Handle the exception appropriately
+            final var rootCause = getRootCause(e);
+            logger.error(String.format("NonRetryable Error: %s", rootCause));
+            throw new NonRetryableException("AlphabeticalIndexUpdaterService.processMessage: ", rootCause);
```
Comment (tombradley1) [inline src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/util/ApiClientUtils.java]: given this is now a generic method in a util class, "AlphabeticalIndexUpdaterService" is no longer appropriate. Perhaps something more generic?

File: src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/service/ResponseHandler.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+import uk.gov.companieshouse.logging.LoggerFactory;
+import uk.gov.companieshouse.alphabeticalcompanysearchconsumer.exception.NonRetryableException;
+import uk.gov.companieshouse.alphabeticalcompanysearchconsumer.exception.RetryableException;
+import static uk.gov.companieshouse.alphabeticalcompanysearchconsumer.logging.LoggingUtils.getLogMap;
+
+
+@Component
+public class ResponseHandler {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(NAMESPACE);
```
Comment (tombradley1) [inline src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/service/ResponseHandler.java]: keep this consistent with the rest of the app; lower case the logger object

**Review by tombradley1 — DISMISSED** [review]

File: src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/util/ApiClientUtils.java (hunk 8 above)
Comment (aokunade) [reply src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/util/ApiClientUtils.java]: MessageProcessingUtil?

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

**Review by tombradley1 — DISMISSED** [review]

File: src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/service/UpsertService.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        try {
+            apiClientService
+                    .getInternalApiClient()
+                    .privateSearchResourceHandler()
+                    .alphabeticalCompanySearch()
+                    .put(resourceUri, companyProfileApi)
+                    .execute();
+            System.out.println("execute successful");
+
+            // return response;
```
Comment (tombradley1) [inline src/main/java/uk/gov/companieshouse/alphabeticalcompanysearchconsumer/service/UpsertService.java]: unnecessary

**Review by tombradley1 — APPROVED** [review]

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

**Pull request merged**